### ISSUE TITLE
Remap keys to mouse buttons 6 to 20

### ISSUE THREAD

**Issue by RamonUnch** [RamonUnch/AltSnap#175]
The window API is limited to 5 mouse buttons (3 buttons +2 XButtons). It is possible using the deprecated DirectInput to access to more than 5 buttons, however it would be quite some work to implement in AltSnap for few people and maybe DInput will not even exist anymore in Win12. So I am not really willing to invest much time in this direction.
However through driver or keymap remapper it is perfectly possible for people with many mouse buttons to bind those buttons to some keybaord keys.

The typical 105-keys keyboard does not have all the possible [key-codes](https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes), for example F13 through F24 (7C-87) typically are not physically present thus can be used without interaction via the mouse buttons.

So I added a few options in AltSnap:
1) you can ask AltSnap to consider some keys as if there where mouse buttons via the `XXButtons` option (see below). Those keys however will be reserved to AltSnap, and will not be usable outside AltSnap.
2) you can now bind Actions for Mouse button 6 to 17 usual `B` suffix (for alternate action) and `T` suffix for Title bar actions do apply too.
3) You can of course use those extra buttons as Hotclicks codes: (07, 08, 09 0A) for 6, 7, 8, 9 (just add one to button number and convert in hexadecimal).

`XXButtons` is a list of keys (hex codes here), 4 keys maximum, the key order matters as the First will correspond to the MB6 the second MB7 etc.
for example:
```
in the [Input] section of the .ini file
XXButtons=7C 7D 7E 7F 80 81 82 83 84 85 86 87
; List of keys that are actually mapped to extra mouse button (6-17)
; You can specify up to 12 values in the button order you prefer.
; Any key specified here will behave like a mouse button for AltSnap.
; Also the key will ALWAYS be unusable outside of AltSnap
; This is intended for mouse buttons that that are mapped by a
; driver to keyboard keys, If possible use keys that does not exist
; on your keyboard ie: F13-F24.
; The extra buttons defined like this are mapped to MB6/7/8/9(B)(T)
; If you set for example XXButtons=7C 7D 7E 7F, then:
; the F13 key will be mapped to MB6, F14 to MB7 etc.
; MB17 is the maximum button. Those buttons can be added to the Hotclicks
```
Now you just need to map your mouse buttons 6, 7, 8,...17 to F13, F14, F15, ..., F24, and you will be able to bind actions to MB6=action1, MB6B=action2, MB6T=action3 etc...

```
Example to use MB6 and 7 for Move resize:
In the [Input] section of the .ini file

MB6=Move
MB7=Resize
; sets up some actions for mouse buttons 6 and 7 

Hotclicks=07 08
; Pressing button 6 or 7 instantly activates AltSnap.
; Values: MMB=04, MB4=05, MB5=06, MB6=07, MB7=08, MB8=09, MB9=0A,
; MB10=0B, MB11=0C, MB12=0D, MB13=0E, MB14=0F, MB15=10,
; MB16=11, MB17=12


XXButtons=7C 7D
; F13 => MB6
; F14 => MB7

```

So the limitation is that the selected keys (in this case F13, F14) are reserved to AltSnap, so just never use keys that you might need elsewhere.
@ariadesu you may be interested.

[hooks1.53test2_i386.zip](https://github.com/RamonUnch/AltSnap/files/8168605/hooks1.53test2_i386.zip)
[hooks1.53test2_x64.zip](https://github.com/RamonUnch/AltSnap/files/8168607/hooks1.53test2_x64.zip)

**Comment by RamonUnch** [comment]
Update: Now you can map up to MB17

**Comment by ariadesu** [comment]
That was quick! That is excellent, thank you for the hard work.

**Comment by RamonUnch** [comment]
I am going to buy a 12 button mouse so I am able to actually test all of this, because for now it is mainly guesswork.

**Comment by RamonUnch** [comment]
I got a new Zelotes C-12 mouse, here is how I set it up:
![12butto_example](https://user-images.githubusercontent.com/74856804/156644344-f7ee021e-6702-49b3-8a54-141c15df90de.png)

With that I set `XXButtons=B5 B1 B0 B3` and now all buttons are mapped correctly.
Of course the drawback is that I can no longer use those multimedia keys on my keyboard but I don't have them anyway, so in my case I really don't care.
It is a shame that the mouse driver does not permit to map mouse buttons to any key, it is missing the F13 to F24 keys like for you.
I was looking to [SharpsKey](https://github.com/randyrants/sharpkeys) in order to remap keys from the mouse (that is actually also a keyboard), but it seems it cannot remap keys for only one of the plugged keyboards, or maybe I am wrong. It seems to be the same for PowerToys' key remapper.

Probably I should use something like [Interception](http://www.oblita.com/interception), [git](https://github.com/oblitum/Interception). to have a finer control. also There is a problem with auto-repeat, it should not apply to mouse buttons, but it is easy to fix.

**Comment by RamonUnch** [comment]
here is an improved version:
1) it goes up to button 20 (I saw there are some 20 mouse buttons around here and the max number of hotkeys in a list is 15 so 15xxbuttons+5base buttons=20 buttons total. I do not think larger numbers would be useful to many people.
2) The auto-repeat is ignored so on some actions this will avoid bugs, when keeping the keys pressed for a while.

[hooks1.53test12_i386.zip](https://github.com/RamonUnch/AltSnap/files/8192273/hooks1.53test12_i386.zip)
[hooks1.53test12_x64.zip](https://github.com/RamonUnch/AltSnap/files/8192274/hooks1.53test12_x64.zip)


It also seems that using PowerToys' key remapper you can remap to **undefined** [**virtual-key codes**](https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes) such as `3A-40` range and `88-8F` range and also `D8-DA` plus a few others. more details [here](https://github.com/microsoft/PowerToys/issues/7544).
This codes can then be used by AltSnap and this way you will be sure it does not interact with anything else.

**Comment by RamonUnch** [comment]
Now available in 1.53 release.
Still a bit *experimental*
Open new issues in case of problems.
